src/Framework/AspNetCoreAnalyzers/test/Http/HeaderDictionaryIndexerAnalyzerTests.cs

```diff
@@ -0,0 +1,152 @@
+
+namespace Microsoft.AspNetCore.Analyzers.Http;
+
+public class HeaderDictionaryIndexerAnalyzerTests
+{
+    [Fact]
+    public async Task IHeaderDictionary_Get_MismatchCase_ReturnDiagnostic()
+    {
+        // Arrange & Act & Assert
+        await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Builder;
+var webApp = WebApplication.Create();
+webApp.Use(async (HttpContext context, Func<Task> next) =>
+{
+    var s = {|#0:context.Request.Headers[""content-type""]|};
+    await next();
+});
+",
+        new DiagnosticResult(DiagnosticDescriptors.UseHeaderDictionaryPropertiesInsteadOfIndexer)
+            .WithLocation(0)
+            .WithMessage("The header 'content-type' can be accessed using the ContentType property"));
+    }
+
+    [Fact]
+    public async Task IHeaderDictionary_Set_MismatchCase_ReturnDiagnostic()
+    {
+        // Arrange & Act & Assert
+        await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Builder;
+var webApp = WebApplication.Create();
+webApp.Use(async (HttpContext context, Func<Task> next) =>
+{
+    {|#0:context.Request.Headers[""content-type""]|} = """";
+    await next();
+});
+",
+        new DiagnosticResult(DiagnosticDescriptors.UseHeaderDictionaryPropertiesInsteadOfIndexer)
+            .WithLocation(0)
+            .WithMessage("The header 'content-type' can be accessed using the ContentType property"));
+    }
+
+    [Fact]
+    public async Task IHeaderDictionary_Get_UnknownProperty_NoDiagnostic()
+    {
+        // Arrange & Act & Assert
+        await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Builder;
+var webApp = WebApplication.Create();
+webApp.Use(async (HttpContext context, Func<Task> next) =>
+{
+    context.Request.Headers[""content-type1""] = """";
+    await next();
+});
+");
+    }
+
+    [Fact]
+    public async Task IHeaderDictionary_Get_VariableProperty_NoDiagnostic()
+    {
+        // Arrange & Act & Assert
+        await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Builder;
+var webApp = WebApplication.Create();
+webApp.Use(async (HttpContext context, Func<Task> next) =>
+{
+    var s = """";
+    context.Request.Headers[s] = """";
+    await next();
+});
+");
+    }
+
+    [Fact]
+    public async Task HeaderDictionary_Get_KnownProperty_NoDiagnostic()
+    {
+        // Arrange & Act & Assert
+        await VerifyCS.VerifyAnalyzerAsync(@"
+using Microsoft.AspNetCore.Http;
+var headers = new HeaderDictionary();
+headers[""Content-Type""] = """";
+");
+    }
+
+    [Fact]
+    public async Task HeaderDictionary_CastToIHeaderDictionary_Get_KnownProperty_ReturnDiagnostic()
```

Comment:
For thoroughness, can we add a test that verifies the analyzers don't emit diagnostics for references that already use DIM properties?

Reply:
Done